[CLI] chia keys derive [wallet-address | child-key | search]

paninaro opened the pull request:
CLI support for displaying or searching derived keys/wallet-addresses. Requires an existing key's fingerprint, or a mnemonic seed to be provided (via a file).

Keys may be derived along a provided HD path, or by specifying a key type `(farmer, pool, wallet, local, backup, singleton, pool_auth)` in which case the HD path is inferred. When specifying an HD path, non-observer derivation can be specified at a given index by appending 'n' to the index. e.g. `m/12381n/8444n/2n/` or `m/12381n/8444n/2/`. If an index in the path does not have the 'n' suffix, observer derivation is assumed at that index. HD paths can mix observer and non-observer derivation indices. Keys derived under the HD path root will use observer derivation by default, or can use non-observer derivation if the `--non-observer-derivation` option is specified.

### Examples

**Show first 10 wallet addresses**
```
$ chia keys derive -f 2543296425 wallet-address -i 0 -n 10 --show-hd-path
Wallet address 0 (m/12381/8444/2/0): xch1m5prhd57vhsywn50c7xn6pxfp6qaakqppqnsx7y8vnefttrml8kq6hqrhy
Wallet address 1 (m/12381/8444/2/1): xch13ugpk9hy8a2csgktqa2l6q2fnyhha7lar3jneaq2kg64xy0dft2qnze2zg
Wallet address 2 (m/12381/8444/2/2): xch1n924lx7wlpw55n4yy23zq8zx2lldhgrylc2l7cmcpsw9vyrxkc0qhsz8l2
Wallet address 3 (m/12381/8444/2/3): xch16ewc6qxfenyd0gu63reykmjxjacyduzuzcvywt5dng3kncvykx4ss2m2wq
Wallet address 4 (m/12381/8444/2/4): xch1ka9kyw2vcv4uqslzfgjuclqyy4xqcmy7yj2uh385knr43vd88cqs3hqswf
Wallet address 5 (m/12381/8444/2/5): xch18dmfjslkxlpaf04dsnwrx0vtgghq6xcclfegkm6v27ek286hczwsm467l6
Wallet address 6 (m/12381/8444/2/6): xch1ehv248e0ntwjhldncdmje5a2lrf6s3ra94tsr4rur4r8tx0sqrzsgd8g3w
Wallet address 7 (m/12381/8444/2/7): xch1hhak3mwm8u2k6thn9l76c76cq7hprxjvswc7ffl6f0yf5vtff3usukfcpt
Wallet address 8 (m/12381/8444/2/8): xch1hq7vg88kvl7lq864d6jpse63yd0fc0whrceja9ec4hhel8yzqumsun9z7j
Wallet address 9 (m/12381/8444/2/9): xch1p20seves63e939qhekt5v3ajw3csqw7ltqycqvy3wea93f9g2axqd6gms3
```

**Show the first singleton pubkey**
```
$ chia keys derive -f 2543296425 child-key -t singleton
Singleton public key 0: b6f7c89c1049df9a9a4188b6186f8fcf02dc2d132e18cd6c0df4e84ca3b270f654b735dfc0a6434c9a2bad2ea573cfee
```

**Show a pair of public and private keys derived from a mixed observer/non-observer HD path using an imported key's mnemonic seed**
```
$ chia keys derive --mnemonic-seed-filename <(chia keys generate_and_print | sed -n 2p) child-key --derive-from-hd-path 'm/12381n/8444n/2/' --show-private-keys --show-hd-path
Observer public key 0 (m/12381n/8444n/2/0): 8e1fbc147c97b57f6678ea5aef32bab0365d3e21319745a0a6bed260222933bbb1f53e3df74d7724d66158e3e101780e
Observer private key 0 (m/12381n/8444n/2/0): 2378d5f8be3590197ba1ae1e7b1d0d55a4151abb2695dec170fb4236c719b8f7
```

**Show the farmer pubkeys 10-14 using an imported key's mnemonic seed**
```
$ chia keys derive --mnemonic-seed-filename <(chia keys generate_and_print | sed -n 2p) child-key -i 10 -n 5 -t farmer
Farmer public key 10: 8710f2d7c424ee98698af5a4e35c151fcc890d895abe58419c70906e0b50c64bb539095d5521432214ce175e432da1c3
Farmer public key 11: b1cfbc0b88b00abed4a3ae558871b7ed24973e8fc17b8c2245ab394955bd8e974f8b7c13de7de7eafc417372c0f34ee5
Farmer public key 12: 8c1e6b53f0b822036fc68885778914069509020cab382a738d6a0381800daf6749adddb79d0617e6cf07ebf887c8ee3b
Farmer public key 13: 861919cfa89a84a066950bc3b92f75367214379c539e35032b1b38cf03b72cec4c0d29f30746e529c4d297741f583031
Farmer public key 14: a606cefbd628e088477b8d4f493cbb43260cb34a10e3a97618c50d212d334ed34cc99fd11b4db1ba10e4586204479195
```

**Search for a wallet address**
```
$ chia keys derive search -t address -l 100 xch15alqzk9sjnzttpx25mar7zvgfakn7xc0dl57r4vugh32vv8u5fsq8lgzx5
Found wallet address: xch15alqzk9sjnzttpx25mar7zvgfakn7xc0dl57r4vugh32vv8u5fsq8lgzx5 (HD path: m/12381/8444/2/75)
```

### CLI Help
```
Usage: chia keys derive [OPTIONS] COMMAND [ARGS]...

Options:
  -f, --fingerprint INTEGER      Enter the fingerprint of the key you want to
                                 use.

  --mnemonic-seed-filename TEXT  The filename containing the mnemonic seed of
                                 the master key to derive from.

  -h, --help                     Show this message and exit.

Commands:
  child-key       Derive child keys
  search          Search the keyring for one or more matching derived keys or
                  wallet addresses

  wallet-address  Derive wallet receive addresses
```
```
Usage: chia keys derive child-key [OPTIONS]

Options:
  -t, --type [farmer|pool|wallet|local|backup|singleton|pool_auth]
                                  Type of child key to derive
  -p, --derive-from-hd-path TEXT  Derive child keys rooted from a specific HD
                                  path. Indices ending in an 'n' indicate that
                                  non-observer derivation should used at that
                                  index. Example HD path: m/12381n/8444n/2/

  -i, --index INTEGER             Index of the first child key to derive.
                                  Index 0 is the first child key.

  -n, --count INTEGER             Number of child keys to derive, starting at
                                  index.

  -d, --non-observer-derivation   Derive keys using non-observer derivation.
                                  [default: False]

  -s, --show-private-keys         Display derived private keys  [default:
                                  False]

  --show-hd-path                  Show the HD path of the derived wallet
                                  addresses  [default: False]

  -h, --help                      Show this message and exit.
```
```
Usage: chia keys derive wallet-address [OPTIONS]

Options:
  -i, --index INTEGER            Index of the first wallet address to derive.
                                 Index 0 is the first wallet address.

  -n, --count INTEGER            Number of wallet addresses to derive,
                                 starting at index.

  -x, --prefix TEXT              Address prefix (xch for mainnet, txch for
                                 testnet)

  -d, --non-observer-derivation  Derive wallet addresses using non-observer
                                 derivation.  [default: False]

  --show-hd-path                 Show the HD path of the derived wallet
                                 addresses. If non-observer-derivation is
                                 specified, path indices will have an 'n'
                                 suffix.  [default: False]

  -h, --help                     Show this message and exit.
```
```
Usage: chia keys derive search [OPTIONS] [SEARCH_TERMS]...

Options:
  -l, --limit INTEGER             Limit the number of derivations to search
                                  against  [default: 100]

  -d, --non-observer-derivation   Search will be performed against keys
                                  derived using non-observer derivation.
                                  [default: False]

  -P, --show-progress             Show search progress  [default: False]
  -t, --search-type [public_key|private_key|address|all]
                                  Limit the search to include just the
                                  specified types  [default: address,
                                  public_key]

  -p, --derive-from-hd-path TEXT  Search for items derived from a specific HD
                                  path. Indices ending in an 'n' indicate that
                                  non-observer derivation should used at that
                                  index. Example HD path: m/12381n/8444n/2/

  -h, --help                      Show this message and exit.
```

paninaro committed:
Added chia keys derive (wallet-address | child-key) to display derived wallet addresses or public keys

emlowe commented:
Nice, jeff. Another option you could add a "find" or "match" option - someone could put in some random address and this returns if it finds it

eFishCent commented:
This looks great.

paninaro commented:
> If we are deriving wallet addresses, i think we should think about what puzzle you are using. Each wallet type has it's own puzzles and therefore it's own way of generating a puzzle from a public key. perhaps the user does not want a standard wallet but a differnt type of wallet, so maybe in the future we want to allow passing in a wallet type, or wallet ID (and is looked up in wallet). That said we don't want to have a dependency of wallet to be running.
> 
> I would definitely like the ability to derive arbitrary paths, and to specify hardened/unhardened. It's not clear which one is being used here.
> 
> Also, how about seeing private keys as well if a flag is specified?

Great points. I hadn't thought too deeply about different wallet types, but that would certainly be useful.

I'd like to add the ability to derive arbitrary paths, supporting hardened and unhardened indices. The notation for hardened being `'n` seems easy enough to support. I'm not clear on a notation for specifying public or private derivation at particular index, but perhaps that doesn't matter? i.e. would it make sense to have a path that mixes public and private derivation like m'/Priv'/Priv'/Priv'/Pub'/Priv/Pub. As I understand it, the extended Priv' key that corresponds to the last extended Pub' key in that path can be obtained and then used to derive an unhardened Priv child key. Perhaps supporting such paths would be a more advanced feature to follow later on.

Showing the leaf private keys via a flag sounds like a good idea.

Thanks for the feedback!

mariano54 commented:
What do you mean by public/private derivation? Are you referring to hardened/unhardened? If so, then yeah, you should be able to mix those.

If you mean private/public keys, you can only go from private -> public. Each key has a public and private representation. The default should be to print the public, but the user might want to see the private key too. We are assuming the private keys are on the machine right? So we can always derive child private keys.

richardkiss commented:
A comment about UI: you might consider supporting `p` or `h` in addition to using a single quote, since single quote is difficult to deal with for command-line utilities, and supporting `p` or `h` avoids having to explain "quoting a single quote" for use in a command-line.

paninaro committed:
Support for public key (unhardened) derivation

paninaro committed:
Fleshed out the search command. Can perform a substring search on wallet addresses, public, and private keys.
Support for deriving private keys

paninaro committed:
Default to unhardened derivation. Added --hardened-derivation option. HD paths denote hardened derivation with a 'h' suffix at each index.

paninaro committed:
Added --no-progress to reduce output. Exit code indicates search success/failure. Option name cleanup.

paninaro committed:
chia keys derive search can now search for multiple terms

paninaro committed:
chia keys derive search can be restricted to searching pubkeys|privkeys|addresses|all

paninaro committed:
Allow child-key derivation rooted from an arbitrary HD path

paninaro committed:
Search derived keys rooted under an arbitrary HD path

paninaro committed:
LGTM fixes

paninaro committed:
Remove extra master_sk_to_wallet_sk_unhardened definition

paninaro committed:
Cleanup and fixed search output

paninaro committed:
Cleanup and comments

paninaro committed:
Added tests for `chia keys` commands

paninaro committed:
Using @pytest_asyncio.fixture to decorate empty_keyring() as suggested in pytest's output.

paninaro committed:
Set asyncio-mode to auto to address pytest warning

paninaro committed:
See if tests pass without checking exit status

paninaro committed:
Added chia init step to tests that direclty invoke the `cli` function

paninaro committed:
Explicitly add --root-path option to tests that run the `cli` function.

paninaro committed:
Tests fix

paninaro committed:
Re-enable checks against exit_code

paninaro committed:
More fine tuning to address CI failures

paninaro commented:
This is ready for review now @mariano54 @emlowe @richardkiss. In addition to the new `chia keys derive` commands, I've also modified `chia keys sign` to accept an HD path that can mix hardened/unhardened indices. Also removed the global `keychain` instance in `key_funcs.py` and replaced it with `Keychain()` instances (which are lightweight now that KeyringWrapper holds the keyring internals).

Tests have been written for all `chia keys` commands.

richardkiss commented on chia/cmds/keys_funcs.py:
```diff
@@ -107,28 +113,501 @@ def delete(fingerprint: int):
+
+    path = path[1:]  # Skip "m"
+
+    if path[-1] == "":  # remove trailing slash
```
Is there supposed to be a slash here?

mariano54 commented on chia/cmds/keys_funcs.py:
```diff
@@ -107,28 +113,501 @@ def delete(fingerprint: int):
+    from chia.wallet.derive_keys import _derive_path, _derive_path_unhardened
+
+    class DerivationType(Enum):
+        HARDENED = 0
```
We should use Observer and non Observer (and probably change the `h` character to `no` in the HD paths). This was the decision that was made regarding terminology.

richardkiss replied:
Parsing the two letters of `no` might be a bit trickier, maybe just `n`? We can easily support multiple markers for hardened derivation. Where we see `if marker == HARDENED_CHAR:` just write `if marker in HARDENED_CHAR_SET:`. I'd recommend supporting both `n` and `h`, maybe `p` (and maybe `no` if it's not too difficult).

richardkiss replied:
Or maybe it's best to let the stupidly ambiguous `p` – the initial letter of both "public" and "private" – die the death it deserves.

mariano54 reviewed (dismissed):
Looks good apart from the naming thing

richardkiss commented on chia/cmds/keys.py:
```diff
@@ -78,9 +80,9 @@ def delete_cmd(ctx: click.Context, fingerprint: int):
+    from chia.util.keychain import Keychain
 
-    keychain.delete_all_keys()
+    Keychain().delete_all_keys()
```
Why is `Keychain()` any better than `keychain`? They both have implicit global state associated with them. What motivated this change?

paninaro replied:
Truthfully I'm not super happy with this, but it was done to support the tests that were written to cover the `keys` commands.

You're right that the `Keychain` wraps some global state held by `KeyringWrapper`. The individual tests will construct a temporary backing keyring, which gets stuffed into the global KeyringWrapper. This presented an issue with the existing `keyring` global instance which held a separate KeyringWrapper instance that wasn't configured properly for testing. The telltale symptom was that when running the tests, the "(Unlock Keyring) Passphrase:" prompt would appear, blocking the tests. So really this was done to ensure that the tests could setup the KeyringWrapper *before* keys.py created its global keychain.

Outside of the test environments, creating a global keychain or many Keychain() instances is effectively the same, minus the cost to construct the objects (which is pretty lightweight at this point).

richardkiss commented on chia/cmds/keys_funcs.py:
```diff
@@ -107,28 +113,501 @@ def delete(fingerprint: int):
+    return AugSchemeMPL.key_gen(seed)
+
+
+def resolve_derivation_master_key(fingerprint: Optional[int], filename: Optional[str]) -> PrivateKey:
```
I think what you really want here is a `Union` type.

`fingerprint_or_filename: Union[int, str]`

and then just use `if isinstance(fingerprint_or_filename, int):` etc.

paninaro replied:
Sure I can make that change. Passing `None` for both params actually is a handled scenario -- the user will be prompted to select a fingerprint from the available keys. I can check `isinstance(fingerprint_or_filename, str)` to catch the filename case, which is the primary condition I want to check.

richardkiss reviewed (dismissed):
I'm approving this anyway because it's so old, and you might understandably be sick of revising it.

paninaro committed:
Changes to address naming and other feedback

paninaro committed:
Updated tests
